adal/src/main/java/com/microsoft/aad/adal/Oauth2.java

```diff
@@ -452,7 +470,26 @@ private AuthenticationResult postMessage(String requestMessage, Map<String, Stri
                 // Protocol related errors will read the error stream and report
                 // the error and error description
                 Logger.v(TAG, "Token request does not have exception");
-                result = processTokenResponse(response);
+                try {
+                    result = processTokenResponse(response);
+                } catch (final AuthenticationException e) {
+                    if (e.getCode().equals(ADALError.SERVER_ERROR_FOR_RETRY) && retryOnce) {
+                        //retry once if it is a server error
+                        //500, 503 and 504 are the ones we retry
+                        retryOnce = false;
+                        try {
+                            //retry once after half second
+                            final CountDownLatch signal = new CountDownLatch(1);
+                            signal.await(delayTimePeriod, TimeUnit.MILLISECONDS);
```

Comment:
Just out of curiosity, what is the benefit of using countdownlatch instead of thread.sleep?